### PR TITLE
Fix possible deadlock when activity resumed during SDK background init

### PR DESCRIPTION
**Pull request by MatkovIvan**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [x] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Android SDK can deadlock with distribute while opening browser if the SDK background thread is initializing. The cause is code using `AppCenter.getInstallId.get()` inside `DistributeUtils` to open browser instead of using asynchronous version (`thenAccept`).

## Related PRs or issues

[AB#74885](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/74885)

### CODE REVIEW

**Comment by MatkovIvan** [comment]
TODOs:

- Update javadocs
- Fix tests and coverage
- Testing via test app
- Changelog

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java [hunk 1]
```diff
@@ -129,23 +141,31 @@ static void updateSetupUsingBrowser(Activity activity, String installUrl, String
...
+        urlBuilder.append("&" + PARAMETER_REDIRECT_ID + "=").append(activity.getPackageName());
+        urlBuilder.append("&" + PARAMETER_REDIRECT_SCHEME + "=" + "appcenter");
+        urlBuilder.append("&" + PARAMETER_REQUEST_ID + "=").append(requestId);
+        urlBuilder.append("&" + PARAMETER_PLATFORM + "=" + PARAMETER_PLATFORM_VALUE);
+        urlBuilder.append("&" + PARAMETER_ENABLE_UPDATE_SETUP_FAILURE_REDIRECT_KEY + "=" + "true");
+        AppCenter.getInstallId().thenAccept(new AppCenterConsumer<UUID>() {
+
+            @Override
+            public void accept(UUID uuid) {
+                urlBuilder.append("&" + PARAMETER_INSTALL_ID + "=").append(uuid.toString());
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java]: state change check

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,12 +2,17 @@
 
 ## Version 2.5.2 (Under development)
 
+### App Center Crashes
+
+* **[Fix]** Fix incorrect app version when an NDK crash is sent after updating the app.
```
Comment (guperrot) [inline CHANGELOG.md]: Please make another PR to fix it now.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 3]
```diff
@@ -968,33 +963,42 @@ synchronized void storeRedirectionParameters(@NonNull String requestId, @NonNull
...
-     * Get latest release details from server.
+     * Request latest release details from server.
      *
      * @param distributionGroupId distribution group id.
      * @param updateToken         token to secure API call.
      */
     @VisibleForTesting
-    synchronized void getLatestReleaseDetails(String distributionGroupId, String updateToken) {
-        AppCenterLog.debug(LOG_TAG, "Get latest release details...");
+    synchronized void requestLatestReleaseDetails(String distributionGroupId, final String updateToken) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Can you explain all these changes about the get API?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 4]
```diff
@@ -1124,7 +1128,7 @@ private synchronized void handleApiCallSuccess(Object releaseCallId, String rawR
 
                 /* Check version code is equals or higher and hash is different. */
                 AppCenterLog.debug(LOG_TAG, "Check if latest release is more recent.");
-                if (isMoreRecent(releaseDetails) && canUpdateNow(releaseDetails)) {
+                if (isMoreRecent(mPackageInfo, releaseDetails) && canUpdateNow(releaseDetails)) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: why passing a member field to a member method parameter?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 5]
```diff
@@ -1231,7 +1203,7 @@ private String getReportingParametersForUpdatedRelease(boolean isPublic, String
     private void changeDistributionGroupIdAfterAppUpdateIfNeeded() {
         String lastDownloadedReleaseHash = SharedPreferencesManager.getString(PREFERENCE_KEY_DOWNLOADED_RELEASE_HASH);
         String lastDownloadedDistributionGroupId = SharedPreferencesManager.getString(PREFERENCE_KEY_DOWNLOADED_DISTRIBUTION_GROUP_ID);
-        if (!isCurrentReleaseWasUpdated(lastDownloadedReleaseHash) || TextUtils.isEmpty(lastDownloadedDistributionGroupId)) {
+        if (!isCurrentReleaseWasUpdated(mPackageInfo, lastDownloadedReleaseHash) || TextUtils.isEmpty(lastDownloadedDistributionGroupId)) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: same here for parameter

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java [hunk 6]
```diff
@@ -129,23 +141,31 @@ static void updateSetupUsingBrowser(Activity activity, String installUrl, String
...
+                String url = urlBuilder.toString();
+                AppCenterLog.debug(LOG_TAG, "No token, need to open browser to url=" + url);
+
+
+                /* Open browser, remember that whatever the outcome to avoid opening it twice. */
+                BrowserUtils.openBrowser(url, activity);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java]: Use the application context and flag new task, since the API is async the activity might be finishing or finished after callback.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 3 above)
Comment (MatkovIvan) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: I need to wrap the call into separate method to just call it in callback. Renaming `get` -> `request` is to avoid confusions because it's not a getter.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 4 above)
Comment (MatkovIvan) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: it's static utils method now. I moved utils method with checks and url building to utils since they do not depend on syncs or many fields. They are good fit for utils class:

https://github.com/microsoft/appcenter-sdk-android/blob/69306570845e0a25db23017bb498178d673a50a0/sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java#L40-L42

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 5 above)
Comment (MatkovIvan) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: it's static method from utils class now

File: CHANGELOG.md (hunk 2 above)
Comment (MatkovIvan) [reply CHANGELOG.md]: #1347
